### PR TITLE
Proof of concept for API and Graphs.jl implementation

### PR DESCRIPTION
**Pull request by GearsAD**

### CODE REVIEW

File: src/entities/DistributedFactorGraph.jl [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import Base: show
+
+struct DistributedFactorGraph
+    # A cache - TBD
+    # something::T
+    # An API for communication
+    api::DFGAPI
+    # A series of mirrors that are asynchronously updated.
+    mirrors::Vector{DFGAPI}
```
Comment (dehann) [inline src/entities/DistributedFactorGraph.jl]: Does this relate to the mirroring happens inside or outside DFG conversation?  I’m nervous about building too much automation into DFG, vs DFG sets the API, establishes symmetry, provides set operations, and lays the ground work for “operating system” features.  Also need to understand how much serialization and deserialization needs to occur in DFG, my guess is at least some would have to occur here and the rest through dispatch to used defined converters.  If you imply mirroring/duplication, I’m just concerned about too much.  Goes back to discussion that duplication happens above DFG in either RoME or Caesar?

File: src/services/GraphsDFG.jl [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+    end
+    return true
+end
+
+"""
+    $(SIGNATURES)
+Get a DFGVariable from a DFG using its underlying integer ID.
+"""
+function getVariable(dfg::GraphsDFG, variableId::Int64)::DFGVariable
+    @warn "This may be slow, rather use by getVariable(dfg, label)"
```
Comment (dehann) [inline src/services/GraphsDFG.jl]: Is this from test results?  My very brief experience was that Symbol lookups on a dict where much much slower than Int key lookups, and strong lookups were much slower than Symbol.  We should confirm.

File: src/services/GraphsDFG.jl [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+
+"""
+    $(SIGNATURES)
+Update a complete DFGVariable in the DFG.
+"""
+function updateVariable!(dfg::GraphsDFG, variable::DFGVariable)::DFGVariable
+    if !haskey(dfg.labelDict, variable.label)
+        error("Variable label '$(variable.label)' does not exist in the factor graph")
+    end
+    dfg.g.vertices[dfg.labelDict[variable.label]].dfgNode = variable
```
Comment (dehann) [inline src/services/GraphsDFG.jl]: I had cases where I ably want to update some of the fields, more a user thing and I’m not exactly sure of how the new use case will work, but just keep in mind till we have completed integration.

File: src/services/GraphsDFG.jl [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+
+"""
+    $(SIGNATURES)
+Delete a DFGVariable from the DFG using its label.
+"""
+function deleteVariable!(dfg::GraphsDFG, label::Symbol)::DFGVariable
+    if !haskey(dfg.labelDict, label)
+        error("Variable label '$(label)' does not exist in the factor graph")
+    end
+    @error "Delete is not supported in Graphs.jl implementation at present"
```
Comment (dehann) [inline src/services/GraphsDFG.jl]: I will add support for delete in graphs.jl

File: src/services/GraphsDFG.jl [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+List the DFGFactors in the DFG.
+Optionally specify a label regular expression to retrieves a subset of the factors.
+"""
+getFactors(dfg::GraphsDFG, regexFilter::Union{Nothing, Regex}=nothing)::Vector{DFGFactor} = lsf(dfg, regexFilter)
+
+"""
+    $(SIGNATURES)
+Checks if the graph is fully connected, returns true if so.
+"""
+function isFullyConnected(dfg::GraphsDFG)::Bool
```
Comment (dehann) [inline src/services/GraphsDFG.jl]: We should remember to add a ‘hasOrphans’ alias

File: src/entities/DistributedFactorGraph.jl (hunk 1 above)
Comment (GearsAD) [reply src/entities/DistributedFactorGraph.jl]: This stuff is older code, saving it for moving it out later. Right now it's inert code though

File: src/services/GraphsDFG.jl (hunk 2 above)
Comment (GearsAD) [reply src/services/GraphsDFG.jl]: The worry isn't actually on the lookup itself, rather on the check to see if the variable is there. Probably O(n) to check if it's in the values of the dictionary. If this is used, then it's probably a good idea to have a idDict
```
    if !(variableId in values(dfg.labelDict))
        error("Variable ID '$(variableId)' does not exist in the factor graph")
    end
```

File: src/services/GraphsDFG.jl (hunk 3 above)
Comment (GearsAD) [reply src/services/GraphsDFG.jl]: Yep! That's part of the actual DFGVariable itself though, more on that when adding the fields into the variables/factors.

File: src/services/GraphsDFG.jl (hunk 5 above)
Comment (GearsAD) [reply src/services/GraphsDFG.jl]: Done!

**Pull request merged**